packages/react/src/useSuspenseCache.tsx

```diff
@@ -0,0 +1,142 @@
+'use client'
+
+import { useEffect, useMemo } from 'react'
+import { useRerender } from './hooks'
+
+type Tuple<T = unknown> = T[] | readonly T[]
+type Cache<Key extends string = string> = {
+  promise?: Promise<unknown>
+  key: Key
+  error?: unknown
+  data?: unknown
+  onNotifies: ((...args: unknown[]) => unknown)[]
+}
+type CacheAction = { reset: () => void }
+type SuspenseCache<TData extends unknown = unknown> = CacheAction & { data: TData }
+
+class SuspenseCacheObserver {
+  private cache = new Map<string, Cache>()
+
+  public reset = <TKey extends string>(key?: TKey) => {
+    if (key === undefined || key.length === 0) {
+      this.cache.clear()
+      this.notifyToAttacher()
+      return
+    }
+
+    if (this.cache.has(key)) {
+      this.cache.delete(key)
+    }
+
+    this.notifyToAttacher(key)
+  }
+
+  public clearError = <TKey extends string>(key?: TKey) => {
+    if (key === undefined || key.length === 0) {
+      this.cache.forEach((value, key, map) => {
+        map.set(key, { ...value, promise: undefined, error: undefined })
+      })
+      return
+    }
+
+    const cacheGot = this.cache.get(key)
+    if (cacheGot) {
+      this.cache.set(key, { ...cacheGot, promise: undefined, error: undefined })
+    }
+  }
+
+  public suspend = <TKey extends string, TData extends unknown>(
+    key: TKey,
+    fn: (options: { key: TKey }) => Promise<TData>
+  ): TData => {
+    const cacheGot = this.cache.get(key)
+
+    if (cacheGot?.error) {
+      // eslint-disable-next-line @typescript-eslint/no-throw-literal
+      throw cacheGot.error
+    }
+    if (cacheGot?.data) {
+      return cacheGot.data as TData
+    }
+
+    if (cacheGot?.promise) {
+      // eslint-disable-next-line @typescript-eslint/no-throw-literal
+      throw cacheGot.promise
+    }
+
+    const newCache: Cache<TKey> = {
+      key,
+      promise: fn({ key })
+        .then((data) => {
+          newCache.data = data
+        })
+        .catch((error) => {
+          newCache.error = error
+        }),
+      onNotifies: [],
+    }
+
+    this.cache.set(key, newCache)
+    // eslint-disable-next-line @typescript-eslint/no-throw-literal
+    throw newCache.promise
+  }
+
+  public getData = <TKey extends string>(key: TKey) => this.cache.get(key)?.data
```

Comment:
what is this function purpose?🤔
it has use case?

Reply:
Oh it is useful😁
hot about this
```suggestion
  public getData = <TKey extends string[]>(key: TKey) => this.cache.get(JSON.stringify(key))?.data
```

because when real use I think it is so hard to use

```tsx
const customCache = useSuspenseCache({
    key: ['custom'],
    fn: () => api.manual({ successPercentage: 50, waitMs: 500 }),
  } as const)

  suspenseCache.getData(JSON.stringify(['custom']))

```

Reply:
So cool!

Reply:
@ChanhyukPark-Tech 
and I edit it as `TKey extends Tuple` not string[] in [28ac54d](https://github.com/suspensive/react/pull/55/commits/28ac54dde7ce65267e699f62d378be3aa49c687c)
I understood your good intention!